### PR TITLE
Better diagnostics for '..' pattern fragment not in the last position

### LINKED ISSUE #49257: Better error message for `..` in struct patterns not in last position
Since the `let P(.., y) = p;` syntax is allowed in unnamed patterns, people will try the same in named patterns, `let P { .., y } = p;`,and it won't work. Right now the error is
```
error: expected `}`, found `,`
   |
13 |     let Point { .., y } = a;
   |                   ^ expected `}`
```
followed by other errors about "pattern does not mention field x". Test code [playground link](https://play.rust-lang.org/?gist=507f4376c637ff7709c11925bc82b6c1&version=nightly)

```rust
struct Point { x: u8, y: u8 }
struct P(u8, u8);

fn main() {
    // Unnamed patterns allow `.., y`
    let p = P(0, 0);
    let P(x, ..) = p;
    let P(.., y) = p;
    
    // But named patterns don't
    let a = Point { x: 0, y: 0 };
    let Point { x, .. } = a; // works
    let Point { .., y } = a; // lots of errors
}
```

Adding a clear error message should be pretty easy, just check if there is a comma after `DotDot` and display a better error message "`..` must always be the last field, and it cannot have a trailing comma". Actually, I think this is the code that checks if `..` is followed by a `}`, so adding the `,` check should be pretty straight-forward:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/libsyntax/parse/parser.rs#L3672-L3677

### PR DESCRIPTION
**Pull request by ordovicia**
Fixes #49257.

### CODE REVIEW

File: src/test/ui/issue-49257.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+#![allow(unused)]
+
+struct Point { x: u8, y: u8 }
+
+fn main() {
+    let p = Point { x: 0, y: 0 };
+    let Point { .., y } = p; //~ ERROR expected `}`, found `,`
+    //~| ERROR pattern does not mention field `x`
+    //~| ERROR pattern does not mention field `y`
```
Comment (petrochenkov) [inline src/test/ui/issue-49257.rs]: Note that the pattern mentions field `y` actually.
Ideally we should do recovery in the parser and continue parsing fields after encountering incorrect `..`, but it requires somewhat larger refactoring of `fn parse_pat_fields`.

Comment (Badel2) [reply src/test/ui/issue-49257.rs]: @petrochenkov Would it be possible to just hide the "pattern does not metion field" errors?

Comment (petrochenkov) [reply src/test/ui/issue-49257.rs]: @Badel2 
Yes, but in this case `Pat { .., field }` or `Pat { field, .., }` have a single obvious meaning, so it's better to use recovery at syntax level and starting from that assume that `Pat { field, .. }` was actually implied.

**Comment by frewsxcv** [comment]
this hit an error in the rollup:

https://travis-ci.org/rust-lang/rust/builds/357368387?utm_source=github_status&utm_medium=notification

@bors r-

**Comment by ordovicia** [comment]
Seems to conflict with #49160.
I'm going to rebase after it is merged (#49308).

**Comment by ordovicia** [comment]
rebased

**Pull request merged**